services/venice-server/src/main/java/com/linkedin/venice/listener/ServerConnectionStatsHandler.java

```diff
@@ -1,74 +1,147 @@
+            serverConnectionStats.incrementRouterConnectionCount();
+          } else {
+            serverConnectionStats.incrementClientConnectionCount();
+          }
+          trackedConnections.add(ctx);
+          newConnections.remove(ctx);
+          Attribute<Boolean> activated = ctx.channel().attr(CHANNEL_ACTIVATED);
+          activated.set(true);
+        }
+      } catch (Exception e) {
+        LOGGER.error("Got exception when scanning new connections", e);
+      }
+    }, 0, 1, java.util.concurrent.TimeUnit.SECONDS);
   }
 
   @Override
   public void channelActive(ChannelHandlerContext ctx) throws Exception {
-    Attribute<Boolean> activated = ctx.channel().attr(CHANNEL_ACTIVATED);
-    if (activated.get() != null && activated.get()) {
-      return;
-    }
-    activated.set(true);
-    SslHandler sslHandler = extractSslHandler(ctx);
-    if (sslHandler == null) {
-      // No ssl enabled, record all connections as client connections
-      serverConnectionStats.incrementClientConnectionCount();
-      return;
-    }
-    String principalName = getPrincipal(sslHandler);
-    if (principalName != null && principalName.contains(routerPrincipalName)) {
-      serverConnectionStats.incrementRouterConnectionCount();
-    } else {
-      serverConnectionStats.incrementClientConnectionCount();
-    }
+    newConnections.add(ctx);
+    super.channelActive(ctx);
   }
 
   @Override
   public void channelInactive(ChannelHandlerContext ctx) throws Exception {
     Attribute<Boolean> activated = ctx.channel().attr(CHANNEL_ACTIVATED);
-    if (activated.get() == null || !activated.get()) {
-      return;
-    }
-    activated.set(false);
-    SslHandler sslHandler = extractSslHandler(ctx);
-    if (sslHandler == null) {
-      // No ssl enabled, record all connections as client connections
-      serverConnectionStats.decrementClientConnectionCount();
-      return;
-    }
-    String principalName = getPrincipal(sslHandler);
-    if (principalName != null && principalName.contains(routerPrincipalName)) {
-      serverConnectionStats.decrementRouterConnectionCount();
-    } else {
-      serverConnectionStats.decrementClientConnectionCount();
+    if (activated.get() != null && activated.get()) {
+      activated.set(false);
+      SslHandler sslHandler = extractSslHandler(ctx);
+      if (sslHandler == null) {
+        LOGGER.error("Failed to extract ssl handler in function: channelInactive");
+      } else {
+        String principalName = getPrincipal(sslHandler);
+        if (principalName == null) {
+          LOGGER.error("Failed to extract principal name from ssl handler in function: channelInactive");
+        } else if (principalName.contains(routerPrincipalName)) {
+          serverConnectionStats.decrementRouterConnectionCount();
+        } else {
+          serverConnectionStats.decrementClientConnectionCount();
+        }
+      }
     }
+
+    newConnections.remove(ctx);
```

Comment:
The `connectionScanner` is scheduled to run every second. In theory there is a chance we might miss processing some connections before it gets removed right? I'm also curious what happens to bad requests? e.g. ACL is bad or we decided to throw error. Do we close the connection right the way?

Reply:
That is correct and if the connection can't be setup properly, such as client cert is bad, it won't be tracked.
For ACL errors, I don't think the connection will be closed and Server would just respond with an error message/code.

One enhancement we can do is to add a new metric to track the connection setup rate regardless of the sources, as in `channelActive` function, it doesn't have access to the client cert right away.